microhapulator/api.py

```diff
@@ -578,61 +579,89 @@ def type(bamfile, markertsv, minbasequal=10, max_depth=1e6):
 
 
 def read_length_dist(
-    fastq,
+    lengthfiles,
     plotfile,
-    lengthsfile=None,
+    samples,
+    hspace=-0.7,
     xlabel="Read Length (bp)",
-    xlim=None,
-    scale=1000,
     title=None,
-    color=None,
-    edgecolor=None,
+    color="red",
+    edgecolor="black",
 ):
     """Plot distribution of read lengths
-
-    :param str fastq: path of a FASTQ file containing NGS reads
+    :param list lengthsfiles: list of JSON files containing read lengths for each sample
     :param str plotfile: path of a graphic file to create
-    :param str lengthsfile: if specified, read lengths will be written to the specified file
     :param str xlabel: label for the X axis
-    :param tuple xlim: a 2-tuple of numbers (x1, x2) representing the start and end points of the portion of the X axis to be displayed; by default this is determined automatically
-    :param float scale: scaling factor for the Y axis
     :param str title: title for the plot
     :param str color: override histogram plot color; red by default
-    :param str edgecolor: override histogram edge color; dark red by default
+    :param str edgecolor: override histogram edge color; black by default
     """
+    read_lengths = aggregate_read_lengths(lengthfiles, samples)
     backend = matplotlib.get_backend()
     plt.switch_backend("Agg")
+    sns.set_theme(style="white", rc={"axes.facecolor": (0, 0, 0, 0), "axes.linewidth": 2})
+    grid = sns.FacetGrid(read_lengths, row="Sample", hue="Sample", aspect=20)
+    grid.map_dataframe(sns.kdeplot, x="ReadLength", color=color, fill=True, alpha=1)
+    grid.map_dataframe(sns.kdeplot, x="ReadLength", color=edgecolor)
+    grid.map(read_distribution_plot_label, "Sample")
+    grid.figure.subplots_adjust(hspace=hspace)
+    minimum_height = 2
+    height = max(minimum_height, len(samples) / 2)
+    grid.figure.set_size_inches(6, height)
+    grid.set_titles("")
+    grid.set(yticks=[], ylabel="")
+    plt.xlabel(xlabel, fontsize=18)
+    plt.xticks(fontsize=18)
+    grid.despine(left=True)
+    if title:
+        plt.title(title, fontsize=18, pad=35)
+    plt.savefig(plotfile, bbox_inches="tight")
+    plt.switch_backend(backend)
```

Comment:
This is the meat of the changes where the plots are actually changed. The other changes are related to only having a single plot with all samples rather than separate plots for each sample.  The values for the spacing parameters were determined by trial and error mostly. In particular `height` -- setting the value based on the number of samples ensures that individual subplots dont' look overly stretched for a low number of samples or overly compressed for a  large number ofo] samples.